flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/KinesisDataFetcher.java

```diff
@@ -182,6 +189,28 @@
 
 	private volatile boolean running = true;
 
+	private final AssignerWithPeriodicWatermarks<T> periodicWatermarkAssigner;
+
+	/**
+	 * The watermark related state for each shard consumer. Entries in this map will be created when shards
+	 * are discovered. After recovery, this shard map will be recreated, possibly with different shard index keys,
+	 * since those are transient and not part of checkpointed state.
+	 */
+	private ConcurrentHashMap<Integer, ShardWatermarkState> shardWatermarks = new ConcurrentHashMap<>();
```

Comment:
So it is not possible to checkpoint this and use it to restore the partition state upon recovery?

Reply:
The mapping can change after recovery. It is not necessary to checkpoint this though because the watermarks are derived information.